feat(refactor): query_slurm function

henrikstranneheim opened the pull request:
## Description

Refactor query_slurm and update_analysis_from_slurm_run_status functions

### Changed

- New utility function to return reformatted SLURM job ids
- New utility function to update analysis status from SLURM run
- Migrate and refactor query_slurm to slurm/api.py
- Extract method from update_analysis_from_slurm_run_status into CRUD
- Remove query_slurm from MockStore
- Add new test fixtures

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b refactor_query_slurm -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

henrikstranneheim commented:
Mere to add pytest-mock to requirements

Vince-janv commented on trailblazer/server/api.py:
```diff
@@ -123,7 +123,7 @@ def update_analysis(analysis_id):
     try:
         process = multiprocessing.Process(
-            target=store.update_run_status, kwargs={"analysis_id": analysis_id, "ssh": True}
+            target=store.update_run_status, kwargs={"analysis_id": analysis_id}
```
Out of curiosity, what happens here? From the API we interact with the database via multiprocessing?

henrikstranneheim replied:
Seems like

henrikstranneheim replied:
Actually slurm on Hasta

Vince-janv commented on trailblazer/store/crud/update.py:
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
+            squeue_response=get_slurm_squeue_output(slurm_job_id_file=Path(analysis.config_path))
+        )
+        self.update_analysis_jobs_from_slurm_jobs(analysis=analysis, squeue_result=squeue_result)
+        LOG.info(f"Status in SLURM: {analysis.family} - {analysis.id}")
```
Is this really needed when we log the new status below?

henrikstranneheim replied:
Changed to debug.

Vince-janv commented on trailblazer/store/api.py:
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
-        analysis: Analysis = self.get_analysis_with_id(analysis_id=analysis_id)
+    def update_analysis_from_slurm_run_status(self, analysis_id: int) -> None:
+        """Query slurm for entries related to given analysis, and update the analysis in the database.
+        Raises: Exception if unable to parse SLURM jobs
```
We don't raise an error in here, we only seem to catch, log and then set the analysis to Error

henrikstranneheim replied:
Removed

Vince-janv commented on tests/cli/test_cli_core.py:
```diff
@@ -146,11 +150,21 @@ def test_cancel_not_running(
+    # GIVEN SLURM squeue output for an analysis
+    mocker.patch(
+        FUNC_GET_SLURM_SQUEUE_OUTPUT_PATH,
+        return_value=subprocess.check_output(
+            ["cat", slurm_squeue_output.get(failed_analysis_case_name)]
+        ).decode("utf-8"),
+    )
```
Neat ⭐

henrikstranneheim replied:
👍 . Was important to move it from the Store

Vince-janv reviewed (approved):
Looks good! 👍

fevac commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,46 @@
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return SLURM squeue jobs flag formatted input."""
+    job_ids: List[str] = slurm_job_id_file_content.get(next(iter(slurm_job_id_file_content)))
```
I find this one liner difficult to read

henrikstranneheim replied:
Modified it.

fevac commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,46 @@
+
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return SLURM squeue jobs flag formatted input."""
```
```suggestion
    """Return a string of comma separated SLURM job ids to be used as input for squeue."""
```

henrikstranneheim replied:
Added

fevac commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,46 @@
+        subprocess.check_output(
+            [
+                "ssh",
+                "hiseq.clinical@hasta.scilifelab.se",
```
with these changes we are always doing ssh to hasta, is that what we want?

henrikstranneheim replied:
No, I misread the code and this would have broken it. Reverted 👍

fevac commented on trailblazer/store/api.py:
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
         if analysis.workflow_manager == WorkflowManager.TOWER.value:
             self.update_tower_run_status(analysis_id=analysis_id)
         elif analysis.workflow_manager == WorkflowManager.SLURM.value:
-            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id, ssh=ssh)
+            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id)
```
do we want to have a controller for this? not sure, up to you, I'm starting to see controllers everywhere lately and I wonder 😬

henrikstranneheim replied:
😄 Not for this

fevac reviewed (commented):
Nice! 🌟 I only got half way. I'll continue tomorrow

fevac commented on trailblazer/store/api.py:
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
-    def update_analysis_from_slurm_run_status(self, analysis_id: int, ssh: bool = False) -> None:
-        """Query slurm for entries related to given analysis, and update the analysis in the database."""
-        analysis: Analysis = self.get_analysis_with_id(analysis_id=analysis_id)
+    def update_analysis_from_slurm_run_status(self, analysis_id: int) -> None:
```
```suggestion
    def update_slurm_run_status(self, analysis_id: int) -> None:
```

henrikstranneheim replied:
But we do not update the slurm run status (we do not control it). We update the analysis based on the current slurm run status.

fevac replied:
you are right. I do find the current name confusing though, the `run_status` part is misleading to me. But I don't have good alternatives to offer, `update_analysis_from_slurm_jobs` ?
 I leave it up to you, feel free to resolve

henrikstranneheim replied:
👍

fevac commented on trailblazer/store/api.py:
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
-        """Query slurm for entries related to given analysis, and update the analysis in the database."""
-        analysis: Analysis = self.get_analysis_with_id(analysis_id=analysis_id)
+    def update_analysis_from_slurm_run_status(self, analysis_id: int) -> None:
+        """Query slurm for entries related to given analysis, and update the analysis in the database.
```
```suggestion
        """Query SLURM for entries related to given analysis, and update the analysis in the database (update jobs, progress, status and logged_at fields).
```

henrikstranneheim replied:
I think this is too specific and easily gets out of sync. I will UC slurm 👍

fevac commented on trailblazer/store/crud/update.py:
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
+        self.update_analysis_jobs_from_slurm_jobs(analysis=analysis, squeue_result=squeue_result)
+        LOG.info(f"Status in SLURM: {analysis.family} - {analysis.id}")
+        LOG.debug(squeue_result.jobs)
+        analysis.progress = squeue_result.jobs_status_distribution.get(
```
type hint

henrikstranneheim replied:
We have decided to not type hint non-self attributes

fevac commented on trailblazer/store/crud/update.py:
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
+        analysis.progress = squeue_result.jobs_status_distribution.get(
+            SlurmJobStatus.COMPLETED, 0.0
+        )
+        analysis.status = get_current_analysis_status(
```
type hint

henrikstranneheim replied:
Same here

fevac commented on trailblazer/store/crud/update.py:
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
+            jobs_status_distribution=squeue_result.jobs_status_distribution
+        )
+        LOG.info(f"Updated status {analysis.family} - {analysis.id}: {analysis.status} ")
+        analysis.logged_at = datetime.now()
```
type hint

fevac commented on tests/cli/test_cli_core.py:
```diff
@@ -457,11 +481,20 @@ def test_scan(
     """Test scanning for analyses and updating analysis status."""
     caplog.set_level("INFO")
 
+    mocker.patch(
+        FUNC_GET_SLURM_SQUEUE_OUTPUT_PATH,
+        return_value=subprocess.check_output(
+            ["cat", slurm_squeue_output.get(ongoing_analysis_case_name)]
+        ).decode("utf-8"),
+    )
+
```
I see this mocker repeated in your tests, is it possible to make a fixture of it? I'm not sure about this suggestion feel free to disregard

henrikstranneheim replied:
I think it is best to use it in test. The  slurm_squeue_output is already a fixture and needs dynamic input from the actual test.

fevac commented on tests/conftest.py:
```diff
@@ -251,3 +256,82 @@ def fixture_tower_task() -> TowerTask:
+        "blazinginsect": Path(
+            "tests",
+            "fixtures",
+            "squeue",
```
can we have a `fixture_directory` so we don't have to repeat this so many times?

henrikstranneheim replied:
Good suggestion - fixed

fevac reviewed (changes requested):
🌟

ChrOertlin commented on tests/conftest.py:
```diff
@@ -251,3 +262,23 @@ def fixture_tower_task() -> TowerTask:
     return TowerTask(**tower_task_running_content["tasks"][0]["task"])
+
+
+@pytest.fixture(scope="session")
+def slurm_squeue_output(squeue_dir: Path) -> Dict[str, Path]:
+    """Return SLURM squeue output for analysis started via SLURM."""
+    file_postfix: str = f"squeue{FileExtension.CSV}"
+    return {
+        "blazinginsect": Path(squeue_dir, f"blazinginsect_{file_postfix}").as_posix(),
+        "crackpanda": Path(squeue_dir, f"crackpanda_{file_postfix}").as_posix(),
+        "cuddlyhen": Path(squeue_dir, f"cuddlyhen_{file_postfix}").as_posix(),
+        "daringpidgeon": Path(squeue_dir, f"daringpidgeon_{file_postfix}").as_posix(),
+        "escapedgoat": Path(squeue_dir, f"escapegoat_{file_postfix}").as_posix(),
+        "fancymole": Path(squeue_dir, f"fancymole_{file_postfix}").as_posix(),
+        "happycow": Path(squeue_dir, f"happycow_{file_postfix}").as_posix(),
+        "lateraligator": Path(squeue_dir, f"lateraligator_{file_postfix}").as_posix(),
+        "liberatedunicorn": Path(squeue_dir, f"liberatedunicorn_{file_postfix}").as_posix(),
+        "nicemice": Path(squeue_dir, f"nicemice_{file_postfix}").as_posix(),
+        "rarekitten": Path(squeue_dir, f"rarekitten_{file_postfix}").as_posix(),
+        "trueferret": Path(squeue_dir, f"trueferret_{file_postfix}").as_posix(),
+    }
```
Perhaps a bit easier on the eyes.
```suggestion
@pytest.fixture(scope="session")
def slurm_squeue_output(squeue_dir: Path) -> Dict[str, Path]:
    """Return SLURM squeue output for analysis started via SLURM."""
    file_postfix: str = f"squeue{FileExtension.CSV}"
    
    job_names = [
        "blazinginsect", "crackpanda", "cuddlyhen", "daringpidgeon", "escapedgoat",
        "fancymole", "happycow", "lateraligator", "liberatedunicorn", "nicemice",
        "rarekitten", "trueferret"
    ]
    
    squeue_output = {}
    
    for job_name in job_names:
        squeue_output[job_name] = Path(squeue_dir, f"{job_name}_{file_postfix}").as_posix()
    
    return squeue_output
```

henrikstranneheim commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,46 @@
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return SLURM squeue jobs flag formatted input."""
+    job_ids: List[str] = slurm_job_id_file_content.get(next(iter(slurm_job_id_file_content)))
```
They are unpacking the dict. Removed for a more explicit approach

ChrOertlin commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,50 @@
+                squeue_commands,
+                universal_newlines=True,
+            )
+            .decode("utf-8")
```
```suggestion
            .decode(UTF8CONSTANT)
```
seen this come around a couple of times as hard coded string in the code

henrikstranneheim replied:
Fixed

ChrOertlin commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,50 @@
+        "%A,%j,%T,%l,%M,%S",
+    ]
+    if use_ssh:
+        squeue_commands = ["ssh", "hiseq.clinical@hasta.scilifelab.se"] + squeue_commands
```
Do we want a constant for the email address?

henrikstranneheim commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,46 @@
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return SLURM squeue jobs flag formatted input."""
+    job_ids: List[str] = slurm_job_id_file_content.get(next(iter(slurm_job_id_file_content)))
+    return ",".join(map(str, job_ids))
```
No, not with updated approch

ChrOertlin reviewed (commented):
Looks good, some style things

henrikstranneheim commented:
@fevac @Vince-janv Ready for another review!

ChrOertlin commented on trailblazer/store/api.py:
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
+    ) -> None:
         """Query slurm for entries related to given analysis, and update the analysis in the database."""
-        analysis: Analysis = self.get_analysis_with_id(analysis_id=analysis_id)
+        analysis: Optional[Analysis] = self.get_analysis_with_id(analysis_id=analysis_id)
```
would checking for 
```Python 
if not analysis:
```
help us define a more accurate error or is the try, except covering this?

henrikstranneheim replied:
It should be handled by the try. Probably need more refactoring to raise more specific exceptions in called functions, but out of scope here.

Vince-janv commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,50 @@
+    return ",".join(job_ids)
+
+
+def get_slurm_squeue_output(slurm_job_id_file: Path, use_ssh: bool = False) -> str:
```
I'm a bit confused regarding the ssh option. What are we using in the automation? I'm not sure we need to support both of we only use one 🤔

henrikstranneheim replied:
FE:
```
process = multiprocessing.Process(target=store.update_ongoing_analyses, kwargs={"use_ssh": True})
       
process = multiprocessing.Process(
            target=store.update_run_status, kwargs={"analysis_id": analysis_id, "use_ssh": True}
```

CLI:
```
context.obj["trailblazer"].update_run_status(analysis_id=analysis_id)
 and several other places
```

So both with and without depending who is calling.
Automation on Hasta do not use ssh

Vince-janv reviewed (approved):
Looks good! Just curious about the ssh option

diitaz93 commented on trailblazer/apps/slurm/api.py:
```diff
@@ -1,10 +1,50 @@
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return a string of comma separated SLURM job ids to be used as input for squeue jobs flag."""
+    job_ids: List[str] = []
+    for slurm_job_ids in slurm_job_id_file_content.values():
+        [job_ids.append(str(job_id)) for job_id in slurm_job_ids]
```
```suggestion
    slurm_job_ids: List[List[str]] = slurm_job_id_file_content.values()
    job_ids: List[str] = [job_id for job_id in job_ids for job_ids in slurm_job_ids]
```

henrikstranneheim replied:
Not a fan of nested for loops within a list comprehension. I think it is hard to understand. I would like to keep the current implementation. Ok?

henrikstranneheim commented on trailblazer/constants.py:
```diff
@@ -75,6 +75,12 @@ def ongoing_statuses(cls) -> Tuple:
 
 
+class CharacterFormat(str, Enum):
+    """Define xharacter encoding/decoding to use."""
```
```suggestion
    """Define character encoding/decoding to use."""
```

fevac commented on trailblazer/server/api.py:
```diff
@@ -113,22 +113,24 @@ def aggregate_jobs():
 @blueprint.route("/update-all")
 def update_analyses():
     """Update all ongoing analysis by querying SLURM"""
```
```suggestion
    """Update all ongoing analysis by querying SLURM."""
```

fevac reviewed (approved):
Very well done! So much nicer ❤️  Looks good to me

diitaz93 commented on tests/store/test_store_api.py:
```diff
@@ -1,13 +1,16 @@
+from trailblazer.constants import CharacterFormat, TrailblazerStatus
 from trailblazer.store.models import Analysis
 
+FUNC_GET_SLURM_SQUEUE_OUTPUT_PATH: str = "trailblazer.store.crud.update.get_slurm_squeue_output"
```
Should this constant be defined only once in a constant file?

henrikstranneheim replied:
It is only for particular tests and I think it makes sense to define it where it is mocked as it has a very special use.

diitaz93 reviewed (commented):
Looks good 👍

henrikstranneheim commented:
### CLI ✅ 

```
θ60° [henrik.stranneheim@hasta:~] [S_base] 2m7s $ trailblazer scan
Status in Tower: richurchin - 278279 - 6231OHy5Yery61
Updated status richurchin - 278279: qc
Updated status mintprawn - 278267: running
Updated status notedocelot - 278263: running
Error updating analysis for: case - finerrhino : PermissionError
Error updating analysis for: case - immuneamoeba : PermissionError
Error updating analysis for: case - modestakita : PermissionError
Error updating analysis for: case - famouscorgi : PermissionError
Error updating analysis for: case - topskite : PermissionError
Error updating analysis for: case - firstviper : PermissionError
Error updating analysis for: case - equalthrush : PermissionError
Error updating analysis for: case - openduck : PermissionError
Error updating analysis for: case - busyturkey : PermissionError
Error updating analysis for: case - strongsponge : PermissionError
Error updating analysis for: case - helpedegret : PermissionError
Error updating analysis for: case - nativeloon : PermissionError
Error updating analysis for: case - supercougar : PermissionError
Error updating analysis for: case - socialcrab : PermissionError
Error updating analysis for: case - activelion : PermissionError
Error updating analysis for: case - amusedspider : PermissionError
Error updating analysis for: case - greatbear : PermissionError
Error updating analysis for: case - aceviper : PermissionError
Error updating analysis for: case - brightape : PermissionError
Error updating analysis for: case - wittyghoul : PermissionError
Error updating analysis for: case - safeshiner : PermissionError
Error updating analysis for: case - brightegret : PermissionError
Error updating analysis for: case - fitloon : PermissionError
Error updating analysis for: case - decentcalf : PermissionError
Error updating analysis for: case - knownburro : PermissionError
Error updating analysis for: case - smoothlocust : PermissionError
Error updating analysis for: case - cuteprawn : PermissionError
Error updating analysis for: case - frankpigeon : PermissionError
Error updating analysis for: case - politeowl : PermissionError
Error updating analysis for: case - likedmutt : PermissionError
Error updating analysis for: case - wholegull : PermissionError
Error updating analysis for: case - promptowl : PermissionError
Error updating analysis for: case - nicewhale : PermissionError
Error updating analysis for: case - activedodo : PermissionError
Error updating analysis for: case - guidedmole : PermissionError
Error updating analysis for: case - busyturtle : PermissionError
Error updating analysis for: case - caringdrum : PermissionError
Error updating analysis for: case - lovedcaiman : PermissionError
Error updating analysis for: case - awakemarten : PermissionError
Error updating analysis for: case - neededbass : PermissionError
Error updating analysis for: case - divinesawfly : PermissionError
Error updating analysis for: case - modestoriole : PermissionError
Error updating analysis for: case - onekoi : PermissionError
Error updating analysis for: case - immunefawn : PermissionError
Error updating analysis for: case - manycow : PermissionError
Error updating analysis for: case - heroicmule : PermissionError
Error updating analysis for: case - inorca : PermissionError
Updated status aceracer - 278157: completed
Error updating analysis for: case - protahr : PermissionError
Error updating analysis for: case - trustysquid : PermissionError
Error updating analysis for: case - solidhusky : PermissionError
Error updating analysis for: case - moralbeagle : PermissionError
Error updating analysis for: case - cracktiger : PermissionError
Error updating analysis for: case - daringrat : PermissionError
Error updating analysis for: case - nativepossum : PermissionError
Error updating analysis for: case - savingstag : FileNotFoundError
Error updating analysis for: case - HW2V2DRX2 : FileNotFoundError
Error updating analysis for: case - divinetomcat : PermissionError
Error updating analysis for: case - handyshrimp : FileNotFoundError
Error updating analysis for: case - setcolt : PermissionError
Error updating analysis for: case - proudfish : PermissionError
Error updating analysis for: case - funnyaphid : FileNotFoundError
Error updating analysis for: case - onestag : PermissionError
Error updating analysis for: case - newgopher : PermissionError
Error updating analysis for: case - helpedlab : MissingSqueueOutput
Error updating analysis for: case - goodbengal : MissingSqueueOutput
Error updating analysis for: case - grandaphid : PermissionError
All analyses updated!
```

### API ✅ 
<img width="816" alt="Screenshot 2023-08-31 at 14 35 34" src="https://github.com/Clinical-Genomics/trailblazer/assets/584738/c9c8e746-f5f2-4b82-90b6-8cf0b863a2fa">

<img width="1426" alt="Screenshot 2023-08-31 at 14 38 18" src="https://github.com/Clinical-Genomics/trailblazer/assets/584738/c241066b-ab8c-4baa-b4d1-18cd17c8df6d">

henrikstranneheim commented:
Deployed in production:
```
Log deploy... done.
trailblazer, version 17.0.7
θ67° [henrik.stranneheim@hasta:~] [P_base] 1m33s $
```
